admin/server/auth/pkce.go

```diff
@@ -0,0 +1,178 @@
+
+	authCodeDB[code] = AuthorizationCode{
+		ClientID:            clientID,
+		RedirectURI:         redirectURI,
+		UserID:              userID,
+		Expiration:          expiration,
+		CodeChallenge:       codeChallenge,
+		CodeChallengeMethod: codeChallengeMethod,
+	}
+
+	// Build the redirection URI with the authorization code as per OAuth2 spec
+	redirectWithCode := fmt.Sprintf("%s?code=%s&state=%s", redirectURI, code, r.URL.Query().Get("state"))
+
+	// Redirect the user agent to the redirect URI with the authorization code
+	http.Redirect(w, r, redirectWithCode, http.StatusFound)
+}
+
+// getAccessTokenForAuthorizationCode exchanges an authorization code for an access token
+func (a *Authenticator) getAccessTokenForAuthorizationCode(w http.ResponseWriter, r *http.Request, values url.Values) {
+	// Extract the authorization code
+	code := values.Get("code")
+	if code == "" {
+		http.Error(w, "authorization code is required", http.StatusBadRequest)
+		return
+	}
+
+	// Extract the client ID
+	clientID := values.Get("client_id")
+	if clientID == "" {
+		http.Error(w, "client ID is required", http.StatusBadRequest)
+		return
+	}
+
+	// Extract the redirect URI
+	redirectURI := values.Get("redirect_uri")
+	if redirectURI == "" {
+		http.Error(w, "redirect URI is required", http.StatusBadRequest)
+		return
+	}
+
+	// Extract the code verifier
+	codeVerifier := values.Get("code_verifier")
+	if codeVerifier == "" {
+		http.Error(w, "code verifier is required", http.StatusBadRequest)
+		return
+	}
+
+	// Validate the authorization code
+	authCode, ok := authCodeDB[code]
+	if !ok {
+		http.Error(w, "invalid authorization code, please re-initiate login", http.StatusBadRequest)
+		return
+	}
+
+	userID := authCode.UserID
+	if userID == "" {
+		http.Error(w, "no user found for authorization code", http.StatusInternalServerError)
+		return
+	}
+
+	// remove the authorization code from the database to prevent reuse
+	delete(authCodeDB, code)
+
+	// Check if the client ID matches the stored client ID
+	if authCode.ClientID != clientID {
+		http.Error(w, "invalid client ID", http.StatusBadRequest)
+		return
+	}
+
+	// Check if the redirect URI matches the stored redirect URI
+	if authCode.RedirectURI != redirectURI {
+		http.Error(w, "invalid redirect URI", http.StatusBadRequest)
+		return
+	}
+
+	// Check if the authorization code has expired
+	if time.Now().After(authCode.Expiration) {
+		http.Error(w, "authorization code has expired", http.StatusBadRequest)
+		return
+	}
+
+	// Verify the code verifier against the stored code challenge
+	if !verifyCodeChallenge(codeVerifier, authCode.CodeChallenge, authCode.CodeChallengeMethod) {
+		http.Error(w, "invalid code verifier", http.StatusBadRequest)
+		return
+	}
+
+	// Issue an access token
+	authToken, err := a.admin.IssueUserAuthToken(r.Context(), userID, authCode.ClientID, "", nil, nil)
+	if err != nil {
+		internalServerError(w, fmt.Errorf("failed to issue access token, %w", err))
+		return
+	}
```

Comment:
If `errors.Is(err, r.Context().Err())`, then we should not return an internal error